torch/distributed/_sharded_tensor/api.py

```diff
@@ -746,3 +923,120 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+    then only stores the local partial shard, process group and the
+    aggregation way to get a full tensor.
+
+    PartialTensor doesn't provide any Tensor like operations but is a
+    wrapper providing the Tensor representing the local partial shard.
+
+    We assume the size of each local tensor to be exactly the same.
+
+    Users can apply custom distributed sharded computations on top of
+    this primitive.
+
+    Args:
+        local_partial_shard (Tensor): Partial result stored across ranks.
+        pg (ProcessGroup): The process group to aggregate on.
+        aggr_op (distributed_c10d.ReduceOp): Way to aggregate the partial result.
+            Default: ``distributed_c10d.ReduceOp.SUM``
+
+    Examples:
+        >>> # All tensors below are of torch.int64 type.
+        >>> # We have 2 process groups, 2 ranks.
+        >>> tensor = torch.arange(2, dtype=torch.int64) + 1 + 2 * rank
+        >>> tensor = torch.stack([tensor, tensor])
+        >>> tensor
+        tensor([[1, 2], [1, 2]]) # Rank 0
+        tensor([[3, 4], [3, 4]]) # Rank 1
+        >>> partial_tensor = _PartialTensor(tensor, distributed_c10d.ReduceOp.MAX)
+        >>> sharding_dim = 0
+        >>> collect_spec = ChunkShardingSpec(
+                dim=sharding_dim,
+                placements=[
+                    "rank:0/cuda:0",
+                    "rank:1/cuda:1",
+                ],
+            )
+        >>> complete_tensor = partial_tensor.reshard(collect_spec)
+        >>> complete_tensor
+        tensor([3, 4]) # Rank 0
+        tensor([3, 4]) # Rank 1
+
+        >>> # All tensors below are of torch.cfloat type.
+        >>> # We have 2 process groups, 2 ranks.
+        >>> tensor = torch.tensor([1+1j, 2+2j], dtype=torch.cfloat) + 2 * rank * (1+1j)
+        >>> tensor = torch.stack([tensor, tensor])
+        >>> tensor
+        tensor([[1.+1.j, 2.+2.j], [1.+1.j, 2.+2.j]]) # Rank 0
+        tensor([[3.+3.j, 4.+4.j], [3.+3.j, 4.+4.j]]) # Rank 1
+        >>> partial_tensor = _PartialTensor(tensor)
+        >>> complete_tensor = partial_tensor.reshard(collect_spec)
+        >>> complete_tensor
+        tensor([4.+4.j, 6.+6.j]) # Rank 0
+        tensor([4.+4.j, 6.+6.j]) # Rank 1
+    """
+
+    def __init__(self, local_partial_shard, pg, aggr_op=distributed_c10d.ReduceOp.SUM):
+        self.local_partial_shard = local_partial_shard
+        self.pg = pg
+        self.aggr_op = aggr_op
+
+    def __post_init__(self):
+        if not isinstance(self.local_partial_shard, torch.Tensor):
+            raise ValueError("local_partial_shard needs to be a Tensor.")
+        if not isinstance(self.aggr_op, distributed_c10d.ReduceOp):
+            raise ValueError(
+                "aggr_op needs to be a member of distributed_c10d.ReduceOp."
+            )
+
+    def reshard(self, resharding_spec: ShardingSpec) -> ShardedTensor:
+        """
+        Consolidate all partial tensors across multiple ranks and covert to a sharded
+        tensor based on the resharding spec.
```

Comment:
Lets add some details here in terms of logically what the result of the operation is:

1. We first aggregate all the shards of the partial tensor.
2. We then shard this tensor according to the provided spec.

Also, add a note that these are logical steps but for performance reasons actual implementation might combine two steps together.

Reply:
Sure.